spring-cloud-gcp-pubsub-stream-binder/src/test/java/org/springframework/cloud/gcp/stream/binder/pubsub/PubSubEmulator.java

```diff
@@ -0,0 +1,216 @@
+					.start();
+		}
+		catch (IOException e) {
+			LOGGER.warn("Gcloud not found; leaving host/port uninitialized");
+			return;
+		}
+
+		if (configPresent) {
+			waitForConfigUpdate(watchService);
+		}
+		else {
+			waitForConfigCreation();
+		}
+
+		Process envInitProcess = new ProcessBuilder("gcloud", "beta", "emulators", "pubsub", "env-init").start();
+
+		// env-init output is a shell command of the form "export
+		// PUBSUB_EMULATOR_HOST=localhost:8085".
+		String emulatorInitString = new BufferedReader(new InputStreamReader(envInitProcess.getInputStream()))
+				.readLine();
+		envInitProcess.waitFor();
+		this.emulatorHostPort = emulatorInitString.substring(emulatorInitString.indexOf('=') + 1);
+	}
+
+	/**
+	 * Shut down the two emulator processes.
+	 *
+	 * gcloud command is shut down through the direct process handle. java process is
+	 * identified and shut down through shell commands.
+	 *
+	 * There should normally be only one process with that host/port combination, but if there
+	 * are more, they will be cleaned up as well.
+	 */
+	@Override
+	protected void after() {
+		if (this.emulatorProcess != null) {
+			this.emulatorProcess.destroy();
+		}
+
+		if (this.emulatorHostPort == null) {
+			return;
+		}
+
+		try {
+			int portSeparatorIndex = this.emulatorHostPort.indexOf(":");
+			if (portSeparatorIndex < 0 || !this.emulatorHostPort.contains("localhost")) {
+				LOGGER.error("Malformed host, can't create emulator: " + this.emulatorHostPort);
+				return;
+			}
+			String emulatorHost = this.emulatorHostPort.substring(0, portSeparatorIndex);
+			String emulatorPort = this.emulatorHostPort.substring(portSeparatorIndex + 1);
+
+			String hostPortParams = String.format("--host=%s --port=%s", emulatorHost, emulatorPort);
+			Process psProcess = new ProcessBuilder("ps", "-v").start();
+			new BufferedReader(new InputStreamReader(psProcess.getInputStream()))
+					.lines()
+					.filter(psLine -> psLine.contains(hostPortParams))
+					.map(psLine -> new StringTokenizer(psLine).nextToken())
+					.forEach(this::killProcess);
+		}
+		catch (IOException e) {
+			LOGGER.error("Failed to cleanup: ", e);
+		}
+	}
+
+	/**
+	 * Return the already-started emulator's host/port combination when called from within a JUnit method.
+	 */
+	public String getEmulatorHostPort() {
+		return this.emulatorHostPort;
+	}
+
+	/**
+	 * Wait until a PubSub emulator configuration file is present.
+	 *
+	 * Fails if the file does not appear after 10 seconds.
+	 *
+	 * @throws InterruptedException which should interrupt the peaceful slumber and bubble up
+	 * to fail the test.
+	 */
+	private void waitForConfigCreation() throws InterruptedException {
+		int attempts = 10;
+		while (!Files.exists(EMULATOR_CONFIG_PATH) && --attempts >= 0) {
+			Thread.sleep(1000);
+		}
+		if (attempts < 0) {
+			throw new RuntimeException(
+					"Emulator could not be configured due to missing env.yaml. Are PubSub and beta tools installed?");
```

Comment:
@artembilan @meltsufin 
Philosophical question -- should this rule never fail, but instead disable the test for any misconfiguration?

Specifically here: if cloud SDK is installed, but cloud beta tools or the pubsub emulator are not installed, should the tests get skipped? Or is the current behavior of failing and telling the user to install beta tools/emulator correct?

Reply:
Well, logically it is not our tests responsibilities that outside software is not configured properly.
We are responsible for the `host/port` only and should fail if there is really no service listening to us.
All other failures of external configuration should only WARN and skip tests.

That's my feeling...